stripe.go

```diff
@@ -487,32 +555,112 @@ func (s *BackendImplementation) Do(req *http.Request, body *bytes.Buffer, v Last
+		// back with status 402.
+		//
+		// On a 402, log to info so as to not make an integration's log
+		// noisy with error messages that they don't have much control
+		// over.
+		//
+		// Note I use the constant 402 instead of an `http.Status*`
+		// constant because technically 402 is "Payment required". The
+		// Stripe API doesn't comply to the letter of the specification
+		// and uses it in a broader sense.
+		if statusCode == 402 {
+			s.LeveledLogger.Infof("User-compelled request error from Stripe (status %v): %v",
+				statusCode, stripeErr.redact())
+		} else {
+			s.LeveledLogger.Errorf("Request error from Stripe (status %v): %v",
+				statusCode, stripeErr.redact())
 		}
+	} else {
+		s.LeveledLogger.Errorf("Error decoding error from Stripe: %v", err)
 	}
+}
+
+// DoStreaming is used by CallStreaming to execute an API request. It uses the
+// backend's HTTP client to execure the request.  In successful cases, it sets
+// a StreamingLastResponse onto v, but in unsuccessful cases handles unmarshaling
+// errors returned by the API.
+func (s *BackendImplementation) DoStreaming(req *http.Request, body *bytes.Buffer, v StreamingLastResponseSetter) error {
+	handleResponse := func(res *http.Response, err error) (interface{}, error) {
+
+		// Some sort of connection error
+		if err != nil {
+			s.LeveledLogger.Errorf("Request failed with error: %v", err)
+			return res.Body, err
+		}
 
+		// Successful response, return the body ReadCloser
+		if res.StatusCode < 400 {
+			return res.Body, err
+		}
+
+		// Failure: try and parse the json of the response
+		// when logging the error
+		var resBody []byte
+		resBody, err = ioutil.ReadAll(res.Body)
+		res.Body.Close()
+		if err == nil {
+			err = s.ResponseToError(res, resBody)
+		} else {
+			s.logError(res.StatusCode, err)
+		}
+
+		return res.Body, err
+	}
+
+	res, result, err := s.requestWithRetriesAndTelemetry(req, body, handleResponse)
 	if err != nil {
 		return err
 	}
+	v.SetLastResponse(newStreamingAPIResponse(res, result.(io.ReadCloser)))
+	return nil
+}
 
-	s.LeveledLogger.Debugf("Response: %s", string(resBody))
+// Do is used by Call to execute an API request and parse the response. It uses
+// the backend's HTTP client to execute the request and unmarshals the response
+// into v. It also handles unmarshaling errors returned by the API.
+func (s *BackendImplementation) Do(req *http.Request, body *bytes.Buffer, v LastResponseSetter) error {
+	handleResponse := func(res *http.Response, err error) (interface{}, error) {
+		var resBody []byte
+		if err == nil {
+			resBody, err = ioutil.ReadAll(res.Body)
+			res.Body.Close()
+		}
+
+		if err != nil {
+			s.LeveledLogger.Errorf("Request failed with error: %v", err)
+		} else if res.StatusCode >= 400 {
+			err = s.ResponseToError(res, resBody)
+
+			s.logError(res.StatusCode, err)
+		}
 
+		return resBody, err
+	}
+
+	res, result, err := s.requestWithRetriesAndTelemetry(req, body, handleResponse)
```

Comment:
Super tiny nit: What do you think of resp instead of res? Having `res` and `result` is a bit confusing to me - `resp` would make it more clear that it's a response.

Reply:
100%, I blame too much express.js for blinding me to the fact that res can mean anything other than response, lol.